### PR TITLE
Clever challenge

### PR DESCRIPTION
**Pull request by padumergue**
## Part 1
In this part I used regex to classify each line of the diff files in the corresponding category and extract the wanted information.

I used go routines to provide diff parsing that scales well the more file there are to parse.
The most difficult part here was with method calls contained inside method calls, and the regex I use might not catch every case.

## Part 2
For this part I worked with the data format of the AST and used recursive tree exploration.
I had to rely on the provided AST specificities to understand how variables declaration are actually represented for primitive types and complex types.

I used the following functions :

extractVariableDeclarations : is the entry point of the tree parsing, it find all trees with VariableDeclaration as root
findNodeWithTypeSequence : to find node that are a the bottom of specific sequence branches
findNodeContainingChildWithType : to find a node that has a specific child
extractValueTextRecursively : to get the succession of values in a subtree

## Part 3
In this part I tried to build a model to predict future sequences based on past observations. Here are the steps I followed. I coded this part using Python3.

### Data preparation : 
I used pandas to open the data source
I tried reducing the dimension of the data using a RandomForestClassifier to select the best features of my events
I also normalized my data before training the model

I splited the events in series that each contain N event to predict the M following events

### Model training :
I choose to use Keras Neural networks and LSTM layers to predict time related data based on articles and tutorials about timeseries prediction.
I fed my model data in shape [n_samples, n_events, n_features] with my y axis with shape [n_samples, n_predicted_classes]
I chose to use binary_crossentropy to train my model as we are trying to classify in classes {0, 1}

### Results :
LSTM did not provide expected results for the classification problem.
When using 10 events to predict the 1 next, the model precision raises slowly on the training set but the validation set stays at around 55% so my model only overfits on the training set.
When looking at the confusion matrix from my predicted classes, I can see that a random classifier would give a result nearly as good on the training set so my model is either badly parametered or not adapted to the problem I'm trying to solve.

As we are trying to classify more than one event in the future I didn't feel like traditional method were adapted but model like HMM would probably give better results.

### Conclusion :
I didn't have enough knowledge about the data I was trying to model and so making assertions was really hard. 
When trying to use the content from res.csv in a one hot encoded matrix I encountered a memory issue and when looking at the used ressources I couldn't find a link between following events.

I could try and tweak my network parameters to see if it provides better results.

### CODE REVIEW

File: sequences/__main__.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    :param y_train: training set class labels
+    :param x_validate: training set attributes values
+    :param y_validate: training set class labels
+    :returns the trained model
+    """
+    model = Sequential()
+    model.add(LSTM(units=64, input_shape=(x_train.shape[1], x_train.shape[2]), kernel_initializer='normal', activation="relu"))
+    model.add(Dense(y_train.shape[1], activation='sigmoid', kernel_initializer='normal'))
+
+    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
```
Comment (chadykamar) [inline sequences/__main__.py:98]: Consider monitoring at precision and recall as well. Accuracy doesn't always the full story.

File: main.go [hunk 2]
```diff
@@ -170,6 +169,113 @@ func extractMethodCalls(line string) *[]string {
...
+	}
+
+	return extractVariables(&ast.Root)
+}
+
+// Explore an AST and find variable declaration
+func extractVariables(node *JsonNode) *astResult {
+	res := astResult{}
+	if node.Type == "VariableDeclaration" {
+		nameNode := findNodeWithTypeSequence(node, []string{"VariableDeclarator", "IdentifierToken"})
```
Comment (mikeoiknine) [inline main.go:207]: If you knew it were possible that for more complicated types, the branches denoting Variable declarations were formed differently, is there a more general way of searching for the appropriate nodes?

File: main.go [hunk 3]
```diff
@@ -170,6 +169,113 @@ func extractMethodCalls(line string) *[]string {
 //computeAST go through the AST and returns
 //a astResult struct that contains all the variable declarations
 func computeAST() *astResult {
+	jsonData, err := ioutil.ReadFile("./ast/astChallenge.json")
+	// if we os.Open returns an error then handle it
+	if err != nil {
+		fmt.Println(err)
```
Comment (mikeoiknine) [inline main.go:188]: This should be handled appropriately

File: main.go [hunk 4]
```diff
@@ -170,6 +169,113 @@ func extractMethodCalls(line string) *[]string {
...
+	}
+
+	var ast Ast
+	err = json.Unmarshal([]byte(jsonData), &ast)
+	if err != nil {
+		fmt.Println("BAD")
```
Comment (mikeoiknine) [inline main.go:194]: Here as well

File: sequences/__main__.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def prepare_data(data):
+    """
+    Prepare the data before training
+    :param data: raw data
+    :return: prepared data
+    """
+    data = data.drop("event_id", axis=1)
+    data = data.drop("timestamp", axis=1)
```
Comment (chadykamar) [inline sequences/__main__.py:41]: Depending on the model, you might want to keep the timestamp but in this case that's fine

File: main.go [hunk 6]
```diff
@@ -42,88 +42,100 @@ func main() {
...
-	}
-
-	scanner := bufio.NewScanner(file)
-
+	// Define all the regex that we need to parse the Diffs
+	var regionPrefix = "^@@"
+	var diffCallPrefix = "diff --git "
+	var diffCallRegex = regexp.MustCompile(regexp.QuoteMeta(diffCallPrefix))
+	var deletedPrefix = "^" + regexp.QuoteMeta("- ")
+	var addedPrefix = "^" + regexp.QuoteMeta("+ ")
```
Comment (mikeoiknine) [inline main.go:88]: I'm not sure I understand why you have an empty character after the `+` and `-`. 

How would you detect something like:

`+void intel_bios_cleanup(struct drm_i915_private *dev_priv);`

File: sequences/__main__.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    model.add(LSTM(units=64, input_shape=(x_train.shape[1], x_train.shape[2]), kernel_initializer='normal', activation="relu"))
+    model.add(Dense(y_train.shape[1], activation='sigmoid', kernel_initializer='normal'))
+
+    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
+    print(model.summary())
+
+    history = model.fit(x_train, y_train, epochs=100, batch_size=64, validation_data=(x_validate, y_validate), verbose=2)
+
+    # plot history
+    pyplot.plot(history.history['loss'], label='train')
```
Comment (chadykamar) [inline sequences/__main__.py:104]: You can use Keras' Tensorboard callback instead

**Review by chadykamar — COMMENTED** [review]
Hello Paul-Antoine,

Thank you for submitting your solution and sorry for not reviewing it sooner.  I looked at the third part and your solution is quite good. You made good use of the various Python libraries, making the code short and simple. The LSTM doesn't have the best results, but it's a very good baseline to build on. Some general nitpicks. 

Some nitpicks:

-  You should have perhaps added a `requirements.txt` file 
- An explanation in the `README` saying that those are the results for 5 timesteps and how to run it for 10, 20 steps

File: sequences/__main__.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def predict_sequence():
+    sample = pd.read_csv("../seq/sample.csv")
+    # ressources = pd.read_csv("../seq/res.csv")
+    sample = prepare_data(sample)
+
+    # create series and split in training and validation sets
+    x_series, y_series = create_series(sample, 5, 1)
+    x_train, x_validate, y_train, y_validate = train_test_split(x_series, y_series, test_size=0.2)
```
Comment (chadykamar) [inline sequences/__main__.py:20]: It's usually best to split the data into 3 parts: training, validation and testing. The validation set can then used to tune the model and monitor the training, reserving the test set just for evaluating the results at the end.

File: main.go [hunk 9]
```diff
@@ -35,13 +42,264 @@ func main() {
...
+
+	return &diffResult
+}
+
+// Parse the content of a diff file and send the result through the channel
+func parseFileDiff(filename string, resChannel chan diffResult) {
+	// Define all the regex that we need to parse the Diffs
+	var regionPrefix = "^@@"
+	var diffCallPrefix = "diff --git "
+	var diffCallRegex = regexp.MustCompile(regexp.QuoteMeta(diffCallPrefix))
```
Comment (lbajolet) [inline main.go:86]: Given that the regexes are static, they should be global as to avoid recompiling them on every use

File: main.go [hunk 10]
```diff
@@ -35,13 +42,264 @@ func main() {
...
+	var deletedPrefix = "^" + regexp.QuoteMeta("- ")
```
Comment (lbajolet) [inline main.go:87]: Generally `regexp.QuoteMeta` can be used to dynamically unquote regexes computed at runtime, for static stuff, I'm not sure of how useful that is, you're probably better off quoting what's problematic. Besides, safe for the `+` one, none actually requires any kind of quoting

File: main.go [hunk 11]
```diff
@@ -35,13 +42,264 @@ func main() {
...
+		functionCalls: make(map[string]int),
+	}
+
+	var matched bool
+	var regexError error
+
+	// Open the file the read line by line
+	file, err := os.Open(filename)
+	if err != nil {
+		log.Fatal(err)
```
Comment (lbajolet) [inline main.go:101]: A log.Fatal looks a bit brutal as the whole program will die when one is executed; this is especially problematic when you have multiple coroutines performing work at the same time, as is your case.

In a cooperative multitasking environment, I'd suggest looking up a way to handle such cases appropriately (error channels is a common way of handling this), though I guess for a simple interview challenge such as this one, this is fine.

File: main.go [hunk 12]
```diff
@@ -35,13 +42,264 @@ func main() {
...
+		expectedResults++
+	}
+
+	diffResult := diffResult{
+		functionCalls: make(map[string]int),
+	}
+
+	// Get results from the communication channel
+	for i := 0; i < expectedResults; i++ {
+		routineRes := <-resChannel
```
Comment (lbajolet) [inline main.go:66]: You should close the channel when you're done with it

File: main.go [hunk 13]
```diff
@@ -35,13 +42,264 @@ func main() {
...
+	err = file.Close()
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// Extract method calls from a given diff line
+func extractMethodCalls(line string) *[]string {
+	// RegEx to find function and method calls
+	functionRegex := regexp.MustCompile("[a-zA-Z_]+" + regexp.QuoteMeta(".") + "?" + "[a-zA-Z]+" + regexp.QuoteMeta("(") + ".*" + regexp.QuoteMeta(")"))
```
Comment (lbajolet) [inline main.go:161]: This will work on `x.y(asd)` but not on `x.y.z(asd)`, and I stand my point, the `QuoteMeta` hinder readability for your regexps
